### ISSUE TITLE
AdWords client is not compatible with perl v5.18

### ISSUE THREAD

**Issue by pickardj79** [googleads/googleads-perl-lib#11]
perl v5.18 changes the behavior of $class->isa. See search.cpan.org/dist/perl-5.18.0/pod/perldelta.pod, search for #47113 and http://www.nntp.perl.org/group/perl.perl5.porters/2012/06/msg188459.html. Google/Ads/SOAP/Typelib/ComplexType.pm, line 87, uses $type->isa('UNIVERSAL') to determine whether or not to load the $type module. Perl 5.18 has changed to return 1 whether or not the $type module is loaded.

This issue can be demonstrated in a couple ways:
1) Demonstrating how perl changed
```
jpickard ~/tmp/AdWords $ perl -v

This is perl 5, version 18, subversion 2 (v5.18.2) built for x86_64-linux

Copyright 1987-2013, Larry Wall

Perl may be copied only under the terms of either the Artistic License or the
GNU General Public License, which may be found in the Perl 5 source kit.

Complete documentation for Perl, including FAQ lists, should be found on
this system using "man perl" or "perldoc perl".  If you have access to the
Internet, point your browser at http://www.perl.org/, the Perl Home Page.

jpickard ~/tmp/AdWords $ perl -e'$type = "Google::Ads::AdWords::v201502::Budget::BudgetDeliveryMethod"; print $type->isa("UNIVERSAL") ? "loaded\n" : "NOT loaded\n"'
loaded
jpickard ~/tmp/AdWords $ export PATH=/usr/local/bin:/usr/bin:/bin:/opt/bin:/usr/x86_64-pc-linux-gnu/gcc-bin/4.6.3:/opt/hadoop/bin:/opt/node/bin
jpickard ~/tmp/AdWords $ perl -v

This is perl 5, version 16, subversion 3 (v5.16.3) built for x86_64-linux
(with 22 registered patches, see perl -V for more detail)

Copyright 1987-2012, Larry Wall

Perl may be copied only under the terms of either the Artistic License or the
GNU General Public License, which may be found in the Perl 5 source kit.

Complete documentation for Perl, including FAQ lists, should be found on
this system using "man perl" or "perldoc perl".  If you have access to the
Internet, point your browser at http://www.perl.org/, the Perl Home Page.

jpickard ~/tmp/AdWords $ perl -e'$type = "Google::Ads::AdWords::v201502::Budget::BudgetDeliveryMethod"; print $type->isa("UNIVERSAL") ? "loaded\n" : "NOT loaded\n"'
NOT loaded
jpickard ~/tmp/AdWords $
```

2) Demonstrating how AdWords code has this issue. v5.18 dies because a module was not loaded. v5.16 dies after that when making a call to Google:
```
jpickard ~/tmp/AdWords $ perl -v

This is perl 5, version 18, subversion 2 (v5.18.2) built for x86_64-linux

Copyright 1987-2013, Larry Wall

Perl may be copied only under the terms of either the Artistic License or the
GNU General Public License, which may be found in the Perl 5 source kit.

Complete documentation for Perl, including FAQ lists, should be found on
this system using "man perl" or "perldoc perl".  If you have access to the
Internet, point your browser at http://www.perl.org/, the Perl Home Page.

jpickard ~/tmp/AdWords $ perl googleads-perl-lib/examples/v201502/campaign_management/get_campaigns_by_label.pl
Use of uninitialized value in concatenation (.) or string at /opt/perl5/perls/perl-5.18.2/lib/site_perl/5.18.2/Google/Ads/Common/HTTPTransport.pm line 32.
Can't locate object method "new" via package "Google::Ads::AdWords::v201502::Predicate::Operator" (perhaps you forgot to load "Google::Ads::AdWords::v201502::Predicate::Operator"?) at /opt/perl5/perls/perl-5.18.2/lib/site_perl/5.18.2/Google/Ads/SOAP/Typelib/ComplexType.pm line 119.
jpickard ~/tmp/AdWords $ export PATH=/usr/local/bin:/usr/bin:/bin:/opt/bin:/usr/x86_64-pc-linux-gnu/gcc-bin/4.6.3:/opt/hadoop/bin:/opt/node/bin
jpickard ~/tmp/AdWords $ perl -v

This is perl 5, version 16, subversion 3 (v5.16.3) built for x86_64-linux
(with 22 registered patches, see perl -V for more detail)

Copyright 1987-2012, Larry Wall

Perl may be copied only under the terms of either the Artistic License or the
GNU General Public License, which may be found in the Perl 5 source kit.

Complete documentation for Perl, including FAQ lists, should be found on
this system using "man perl" or "perldoc perl".  If you have access to the
Internet, point your browser at http://www.perl.org/, the Perl Home Page.

jpickard ~/tmp/AdWords $ perl googleads-perl-lib/examples/v201502/campaign_management/get_campaigns_by_label.pl
Use of uninitialized value in concatenation (.) or string at /usr/lib64/perl5/vendor_perl/5.16.3/Google/Ads/Common/HTTPTransport.pm line 32.
[21 May 2015 16:38:07,353 - INFO ] Outgoing Request:
<SOAP-ENV:Envelope xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:SOAP-ENV="http://schemas.xmlsoap.org/soap/envelope/" ><SOAP-ENV:Header><RequestHeader xmlns="https://adwords.google.com/api/adwords/cm/v201502"><userAgent xmlns="https://adwords.google.com/api/adwords/cm/v201502">googleads-perl-lib/examples/v201502/campaign_management/get_campaigns_by_label.pl (AwApi-Perl/3.3.0, Common-Perl/3.1.1, SOAP-WSDL/2.00.10, libwww-perl/6.05, perl/5.016003)</userAgent><validateOnly xmlns="https://adwords.google.com/api/adwords/cm/v201502">false</validateOnly><partialFailure xmlns="https://adwords.google.com/api/adwords/cm/v201502">false</partialFailure></RequestHeader></SOAP-ENV:Header><SOAP-ENV:Body><get xmlns="https://adwords.google.com/api/adwords/cm/v201502"><serviceSelector><fields>Id</fields><fields>Name</fields><fields>Labels</fields><predicates><field>Labels</field><operator>CONTAINS_ANY</operator><values>INSERT_LABEL_ID_HERE</values></predicates><ordering><field>Name</field><sortOrder>ASCENDING</sortOrder></ordering><paging><startIndex>0</startIndex><numberResults>500</numberResults></paging></serviceSelector></get></SOAP-ENV:Body></SOAP-ENV:Envelope>
The library couldn't find any authorization mechanism set up to properly sign the requests against the API. Please read the following guide on how to setup OAuth2 https://github.com/googleads/googleads-perl-lib/wiki/Using-OAuth-2.0 at /usr/lib64/perl5/vendor_perl/5.16.3/Google/Ads/Common/HTTPTransport.pm line 47.
```

**Comment by sundquist** [comment]
We're looking into fixing this issue. Thank you for providing such a thorough and detailed explanation of the issue. It was extremely helpful.

Cheers,
Nadine Sundquist, AdWords API Team

**Comment by sundquist** [comment]
This is now fixed with version 4.0.0.
